### ISSUE TITLE
Dictionary passed into machine class is not used on slack client

### ISSUE THREAD

**Issue by hhollenstain** [DandyDev/slack-machine#390]
# Issue
Passing settings into the Machine class is ignored when generating the SlackClient. The Slack client class just runs a local_settings import via import_settings() when this seems to be unnecessary. 

```python 

from machine import Machine
from machine.utils.collections import CaseInsensitiveDict

base_settings = {
    'PLUGINS': [],
    'STORAGE_BACKEND': 'machine.storage.backends.memory.MemoryStorage',
    'DISABLE_HTTP': True,
    'HTTP_SERVER_HOST': '0.0.0.0',
    'HTTP_SERVER_PORT': 8080,
    'HTTP_SERVER_BACKEND': 'wsgiref',
    'HTTP_PROXY': None,
    'HTTPS_PROXY': None,
    'KEEP_ALIVE': None,
    'SLACK_API_TOKEN': '<redacted token>'
}

settings = CaseInsensitiveDict(base_settings)
bot = Machine(settings=settings)
bot.run()
```
output:

```bash
pipenv run python test-bot.py                                     
Loading .env environment variables…
Initializing Slack Machine:
    Loading settings...
Traceback (most recent call last):
  File "test-bot.py", line 18, in <module>
    bot = Machine(settings=settings)
  File "/Users/henry.hollenstain/.virtualenvs/bot-9dQm3zAV/lib/python3.7/site-packages/machine/core.py", line 59, in __init__
    self._client = LowLevelSlackClient()
  File "/Users/henry.hollenstain/.virtualenvs/bot-9dQm3zAV/lib/python3.7/site-packages/machine/utils/__init__.py", line 14, in __call__
    cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
  File "/Users/henry.hollenstain/.virtualenvs/bot-9dQm3zAV/lib/python3.7/site-packages/machine/clients/singletons/slack.py", line 33, in __init__
    self.rtm_client = RTMClient(token=slack_api_token, proxy=http_proxy)
  File "/Users/henry.hollenstain/.virtualenvs/bot-9dQm3zAV/lib/python3.7/site-packages/slack/rtm/client.py", line 118, in __init__
    self.token = token.strip()
AttributeError: 'NoneType' object has no attribute 'strip'
```

Placing a local_settings.py file with a token works.
```python
SLACK_API_TOKEN = 'redacted' 
```

```bash
pipenv run python test-bot.py                                           
Loading .env environment variables…
Initializing Slack Machine:
    Loading settings...
    Initializing storage using backend: machine.storage.backends.memory.MemoryStorage
    Loading plugins...

Starting Slack Machine:
    ✓ Connected to Slack
    ✓ Scheduler started
    ✓ Dispatcher started
```

It seems when generating the slack client https://github.com/DandyDev/slack-machine/blob/master/machine/core.py#L59

The client re imports settings already on the Machine class instead of passing in the settings to the LowLevelSlackClient: 
https://github.com/DandyDev/slack-machine/blob/master/machine/clients/singletons/slack.py#L30

## Proposed solution
Pass settings from the Machine class into LowLevelSlackClient

**Comment by DandyDev** [comment]
@hhollenstain that is true and earlier this week when I was fixing tests, I discovered this problem as well. Currently, the settings passed into the `Machine` class are only used to determine which plugins to load. Ideally, we would use those settings for instantiating the underlying Slack client and storage adapter as well, but that is not possible: both Slack client and storage adapter are singletons that cannot be instantiated with any parameters because that means I have to remove the singleton aspect from them. 

Why are they singletons (that's a bad software engineering pattern)? Because that's the only way they'll play nice with the scheduler. APScheduler will serialize scheduled jobs such as when you use the `@schedule` decorator. The functions decorated with the `@schedule` decorator are part of classes extending `MachineBasePlugin` and APScheduler will try to serialize the whole object (class) that contains the decorated function. That won't work if the object contains references, even indirectly, to sockets. Both Slack client and for example the Redis storage backend contain sockets. Hence the need to make them singletons. Then I don't have to pass them into the `MachineBasePlugin` when instantiating plugins (which, from a engineering perspective would be the better solution). Instead, `MachineBasePlugin` will just retrieve the only instance in existence from those singletons.

Long story short: not possible, so counterintuitively, it's probably better to remove `settings` from the `Machine` constructor altogether as to not give the impression that those settings will be used for everything.

I'm open to other suggestions though!

**Comment by hhollenstain** [comment]
@DandyDev Thanks for the quick reply and insight. After your explanation I don't have any suggestions at this time. If I get any inspiration as I work around using slack-machine more than I did yesterday I will be circling back here. Again thanks!

**Comment by DandyDev** [comment]
Only after I wrote my reply, I actually saw your PR. With some adjustments it actually might work. I will have a look at your PR in the coming days, so don't close it yet.

**Comment by DandyDev** [comment]
I do have a question though: is there a reason you want to provide the settings to `Machine` directly instead of using a `local_settings.py`?

**Comment by hhollenstain** [comment]
The reason I was looking at populating the settings dictionary prior to starting the machine in code is due to our access pattern of secrets. We utilize vault client to pull and populate things such as tokens from a client. Idea is to seed this data and provide to the Machine class with this seeded data. Again we have options around this such as generating the file prior to starting the Machine, but would be great to not do that. Hope that clears up why I am interested in this, again not a blocker by any stretch.

**Comment by DandyDev** [comment]
Can't you expose secrets as environment variables? Slack Machine accepts those as well. You can supply any setting through environment variables by prefixing them with `SM_`. So `SLACK_TOKEN` becomes `SM_SLACK_TOKEN` for example.

**Comment by hhollenstain** [comment]
@DandyDev  That  is a great idea. I forgot you are checking for SM_. Able to patch the os.environ dict with SM_SLACK_API and worked perfectly.
